File: pkg/tracer/transformations.go
package tracer

import "math"

func TranslationMatrix(x, y, z float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	m[0][3] = x
	m[1][3] = y
	m[2][3] = z
	return m
}

func ScalingMatrix(x, y, z float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	m[0][0] = x
	m[1][1] = y
	m[2][2] = z
	return m
}

func RotateXMatrix(rad float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	sin, cos := math.Sin(rad), math.Cos(rad)
	m[1][1] = cos
	m[1][2] = -sin
	m[2][1] = sin
	m[2][2] = cos
	return m
}

func RotateYMatrix(rad float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	sin, cos := math.Sin(rad), math.Cos(rad)
	m[0][0] = cos
	m[0][2] = sin
	m[2][0] = -sin
	m[2][2] = cos
	return m
}

func RotateZMatrix(rad float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	sin, cos := math.Sin(rad), math.Cos(rad)
	m[0][0] = cos
	m[0][1] = -sin
	m[1][0] = sin
	m[1][1] = cos
	return m
}

func ShearingMatrix(xy, xz, yx, yz, zx, zy float64) *Mat4 {
	m := MatrixIdentity(4).(*Mat4)
	m[0][1] = xy
	m[0][2] = xz
	m[1][0] = yx
	m[1][2] = yz
	m[2][0] = zx
	m[2][1] = zy
	return m
}

func ViewMatrix(from, to, up *Tup) *Mat4 {
	forward := SubTup(to, from).Normalize()
	left := CrossVec(forward, up.Normalize())
	trueUp := CrossVec(left, forward)

	m := MatrixIdentity(4).(*Mat4)

	m[0][0], m[0][1], m[0][2] = left.X, left.Y, left.Z
	m[1][0], m[1][1], m[1][2] = trueUp.X, trueUp.Y, trueUp.Z

	m[2][0], m[2][1], m[2][2] = -forward.X, -forward.Y, -forward.Z

	return MatrixMultiply(m, TranslationMatrix(-from.X, -from.Y, -from.Z))
}

File: src/tracer/transformations.rs
use super::matrix::{Kind, Mat};
use super::tuple;

pub fn translation(x: f32, y: f32, z: f32) -> Mat {
    let mat = [
        [1.0, 0.0, 0.0, x],
        [0.0, 1.0, 0.0, y],
        [0.0, 0.0, 1.0, z],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::TransformNoScale;

    Mat::new(mat, kind)
}

pub fn scaling(x: f32, y: f32, z: f32) -> Mat {
    let mat = [
        [x, 0.0, 0.0, 0.0],
        [0.0, y, 0.0, 0.0],
        [0.0, 0.0, z, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::Transform;

    Mat::new(mat, kind)
}

pub fn rotate_x(rad: f32) -> Mat {
    let sin = rad.sin();
    let cos = rad.cos();
    let mat = [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, cos, -sin, 0.0],
        [0.0, sin, cos, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::TransformNoScale;

    Mat::new(mat, kind)
}

pub fn rotate_y(rad: f32) -> Mat {
    let sin = rad.sin();
    let cos = rad.cos();
    let mat = [
        [cos, 0.0, sin, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [-sin, 0.0, cos, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::TransformNoScale;

    Mat::new(mat, kind)
}

pub fn rotate_z(rad: f32) -> Mat {
    let sin = rad.sin();
    let cos = rad.cos();
    let mat = [
        [cos, -sin, 0.0, 0.0],
        [sin, cos, 0.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::TransformNoScale;

    Mat::new(mat, kind)
}

pub fn shearing(xy: f32, xz: f32, yx: f32, yz: f32, zx: f32, zy: f32) -> Mat {
    let mat = [
        [1.0, xy, xz, 0.0],
        [yx, 1.0, yz, 0.0],
        [zx, zy, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::General;

    Mat::new(mat, kind)
}

pub fn view(from: tuple::Tup, to: tuple::Tup, up: tuple::Tup) -> Mat {
    let forward = (&to - &from).normalize();
    let left = tuple::cross(&forward, &up.normalize());
    let true_up = tuple::cross(&left, &forward);

    let mat = [
        [left.x, left.y, left.z, 0.0],
        [true_up.x, true_up.y, true_up.z, 0.0],
        [-forward.x, -forward.y, -forward.z, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    let kind = Kind::General;

    Mat::new(mat, kind) * translation(-from.x, -from.y, -from.z)
}
